src/AwkwardArray.jl

```diff
@@ -445,4 +718,336 @@ function end_record!(layout::RecordArray)
+    parameters::Parameters = Parameters(),
+    behavior::Symbol = :default,
+) where {INDEX<:Index8} where {CONTENT<:Content} = ByteMaskedArray(
+    INDEX([]),
+    CONTENT(),
+    valid_when = valid_when,
+    parameters = parameters,
+    behavior = behavior,
+)
+
+function copy(
+    layout::ByteMaskedArray{INDEX1,CONTENT1,BEHAVIOR};
+    mask::Union{Unset,INDEX2} = Unset(),
+    content::Union{Unset,CONTENT2} = Unset(),
+    valid_when::Union{Unset,Bool} = Unset(),
+    parameters::Union{Unset,Parameters} = Unset(),
+    behavior::Union{Unset,Symbol} = Unset(),
+) where {INDEX1<:Index8,INDEX2<:Index8,CONTENT1<:Content,CONTENT2<:Content,BEHAVIOR}
+    if isa(mask, Unset)
+        mask = layout.mask
+    end
+    if isa(content, Unset)
+        content = layout.content
+    end
+    if isa(valid_when, Unset)
+        valid_when = layout.valid_when
+    end
+    if isa(parameters, Unset)
+        parameters = parameters_of(layout)
+    end
+    if isa(behavior, Unset)
+        behavior = typeof(layout).parameters[end]
+    end
+    ByteMaskedArray(
+        mask,
+        content,
+        valid_when = valid_when,
+        parameters = parameters,
+        behavior = behavior,
+    )
+end
+
+function is_valid(layout::ByteMaskedArray)
+    if length(layout.mask) > length(layout.content)
+        return false
+    end
+    return is_valid(layout.content)
+end
+
+Base.length(layout::ByteMaskedArray) = length(layout.mask)
+Base.firstindex(layout::ByteMaskedArray) = firstindex(layout.mask)
+Base.lastindex(layout::ByteMaskedArray) = lastindex(layout.mask)
+
+function Base.getindex(layout::ByteMaskedArray, i::Int)
+    if (layout.mask[i] != 0) != layout.valid_when
+        nothing
+    else
+        adjustment = firstindex(layout.mask) - firstindex(layout.content)
+        layout.content[i-adjustment]
+    end
+end
+
+function Base.getindex(layout::ByteMaskedArray, r::UnitRange{Int})
+    adjustment = firstindex(layout.mask) - firstindex(layout.content)
+    copy(
+        layout,
+        mask = layout.mask[r.start:r.stop],
+        content = layout.content[(r.start-adjustment):(r.stop-adjustment)],
+    )
+end
+
+Base.getindex(layout::ByteMaskedArray, f::Symbol) =
+    copy(layout, content = layout.content[f])
+
+function push!(
+    layout::ByteMaskedArray{INDEX,CONTENT},
+    x::ITEM,
+) where {INDEX<:Index8,ITEM,CONTENT<:PrimitiveArray{ITEM}}
+    push!(layout.content, x)
+    Base.push!(layout.mask, layout.valid_when)
+    layout
+end
+
+function end_list!(layout::ByteMaskedArray)
+    end_list!(layout.content)
+    Base.push!(layout.mask, layout.valid_when)
+    layout
+end
+
+function end_record!(layout::ByteMaskedArray)
+    end_record!(layout.content)
+    Base.push!(layout.mask, layout.valid_when)
+    layout
+end
+
+function push_null!(
+    layout::ByteMaskedArray{INDEX,CONTENT},
+) where {INDEX<:Index8,ITEM,CONTENT<:PrimitiveArray{ITEM}}
+    push!(layout.content, ITEM(0))  # FIXME: how to get a value for any ITEM type?
```

Comment:
Here's a puzzle: given a type `ITEM`, I want to get an instance of `ITEM`. I don't care what that instance is—it's going to be hidden behind a mask declaring it to be missing—but I want to create it generically for any type `ITEM`.

The above may be good enough because in Python Awkward Arrays, the only values that can go into a `PrimitiveArray` (`NumpyArray`) are from [this list](https://github.com/scikit-hep/awkward/blob/0accfadd25d6562813e91771177bb5b0bda5a7b1/src/awkward/types/numpytype.py#L66-L93):

  * `Bool`
  * `Int8`, `UInt8`, `Int16`, `UInt16`, `Int32`, `UInt32`, `Int64`, `UInt64`
  * `Float16`, `Float32`, `Float64`
  * `Complex{Float32}`, `Complex{Float64}`
  * `Dates.DateTime`, `Dates.TimePeriod`
  * `Float128` and `Complex{Float128}` if it can be obtained somehow

and all of these have constructors that would take `0` as the only argument. So for any Awkward Arrays that can be shared with Python, `ITEM(0)` is sufficient.

Reply:
ah, `zero(ITEM)` is the idiom
```
julia> zero(Int)
0

julia> zero(Int32)
0

julia> zero(UInt32)
0x00000000
```

Reply:
Awesome! See 7655d8d5ad8d3078a726617db93070a4919a618b.